File: expansion_board/i2c/oled/oled_dht11/python/oled_dht11.py
###############################################
#
#  file: oled_dht11.py
#  update: 2024-10-22
#  usage: 
#      sudo python oled_dht11.py
#
###############################################

import time
import gpiod
import os
import struct
import subprocess
import board
import busio
from PIL import Image, ImageDraw, ImageFont
import adafruit_ssd1306

# oled初始化
i2c = busio.I2C(board.I2C3_SCL, board.I2C3_SDA)
disp = adafruit_ssd1306.SSD1306_I2C(128, 64, i2c)

width = disp.width  
height = disp.height  
image = Image.new('1', (width, height))  
draw = ImageDraw.Draw(image)
draw.rectangle((0, 0, width, height), outline=0, fill=0)

font = ImageFont.truetype("NotoSerifCJKhk-VF.ttf", 16, encoding="unic")  
x = 5   
top = 5  
#-----------------------------------------------------------------

# dht11初始化
MAX_TEMPERATURE = 35    
MAX_HUMIDITY = 60
dht11_fd = os.open('/dev/dht11', os.O_RDONLY)
#-----------------------------------------------------------------

def main():
    try:
        while True:
            
            draw.rectangle((0, 0, width, height), outline=0, fill=0)

            # 获取温湿度值并解析
            data = os.read(dht11_fd, 6)
            temp = float(data[2] + data[3] * 0.01)
            humi = float(data[0] + data[1] * 0.01)
            #print("temp: " + str(temp) + " humi: " + str(humi))

            # 绘制新文本
            draw.text((x, top + 16), "temp: " + str(temp) + "℃", font=font, fill=255)
            draw.text((x, top + 32), "humi: " + str(humi) + "%", font=font, fill=255)

            # oled显示图像
            disp.image(image)  
            disp.show() 
            
            time.sleep(1)

    except Exception as e:
        
        print("exit...：", e)
    
    finally:

        # disp clear
        disp.fill(0)
        disp.show() 

        # dht11 close
        os.close(dht11_fd)
        
if __name__ == "__main__":  
    main()

File: expansion_board/i2c/oled/oled_dht11/c/oled_dht11.c
/*
*
*   file: main.c
*   update: 2024-08-09
*   usage: 
*       sudo make
*       sudo ./oled_dht11
*
*/

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <stdint.h>
#include <signal.h> 

#include "oled.h"

/* dht11设备名 */
#define DEV_NAME    "/dev/dht11"

/* 温湿度阈值 */
#define MAX_TEMP    (35)
#define MAX_HUMI    (60)

int dht11_fd;

static void sigint_handler(int sig_num) 
{    
    /* oled清屏 */
    oled_clear();

    /* 关闭dht11 */
    close(dht11_fd);

    exit(0);  
}

int main(int argc, char **argv)
{
    int ret = 0;
    uint8_t data[6];
    float temp, humi;
    char temp_str[50], humi_str[50];

    /* register exit signal ( Ctrl + c ) */
    signal(SIGINT, sigint_handler);

    /* dht11初始化 */
    dht11_fd = open(DEV_NAME, O_RDWR);
	if(dht11_fd < 0)
	{
		printf("can not open file %s, %d\n", DEV_NAME, dht11_fd);
		return -1;
	}

    /* oled初始化 */
    ret = oled_init(3);
    if(ret == -1)
    {
        printf("oled init err!\n");
        return -1;
    }
    
    while(1)
    {
        /* 读取dht11温湿度数据 */
		ret = read(dht11_fd, &data, sizeof(data));	
		if(ret)
        {
            temp = data[2] + data[3] * 0.01;
            humi = data[0] + data[1] * 0.01;
        }
        else
            printf("read data from dth11 err!\n");

        sprintf(temp_str, "temp: %.2f", temp);
        sprintf(humi_str, "humi: %.2f", humi);
        oled_show_string(0, 2, temp_str);
        oled_show_string(0, 4, humi_str);

        sleep(1);
    }

    close(dht11_fd);

    return 0;
}